README.md

```diff
@@ -58,32 +124,312 @@ client = Redis.new(semian: {
+a bad user experience, and ultimately compromise the performance of the entire
+system. Even if the timeout is as low as ~250ms (just enough to allow a single
+TCP retransmit) there's a large loss of capacity and for many application
+100-300% increase in average response time. This is the problem Semian solves by
+failing fast.
+
+## How does Semian work?
+
+Semian consists of two parts: circuit breaker and bulkheading. To understand
+Semian, and especially how to configure it, we must understand these patterns
+and their implementation. 
+
+### Circuit Breaker
+
+The circuit breaker pattern is based on a simple observation - if we hit a
+timeout for a given service one or more times, we’re likely to hit it again for
+some amount of time. Instead of hitting the timeout repeatedly, we can mark the
+resource as dead for some amount of time during which we raise an exception
+instantly on any call to it. This is called the [circuit breaker pattern][cbp].
+
+![](http://cdn.shopify.com/s/files/1/0070/7032/files/image01_grande.png)
+
+When we do a Remote Procedure Call (RPC), it will first check the circuit. If
+the circuit is rejecting requests because of too many failures reported by the
+driver, it will throw an exception immediately. Otherwise the circuit will call
+the driver. If the driver fails to get data back from the data store, it will
+notify the circuit. The circuit will count the error so that if too many errors
+have happened recently, it will start rejecting requests immediately instead of
+waiting for the driver to time out. The exception will then be thrown back to
+the original caller. If the driver’s request was successful, it will return the
+data back to the calling method and notify the circuit that it made a successful
+call.
+
+The state of the circuit breaker is local to the worker and is not shared across
+all workers on a server.
+
+#### Circuit Breaker Configuration
+
+There are three configuration parameters for circuit breakers in Semian:
+
+* **error_threshold**. The amount of errors to encounter for the worker before
+  opening the circuit, i.e. start rejecting requests instantly.
+* **error_timeout**. The amount of time until trying to query the resource
+  again.
+* **success_threshold**. The amount of successes on the circuit until closing it
+  again. I.e. accept all requests to the circuit.
+
+### Bulkheading
+
+However, circuit breakers for many applications are not enough. This is best
+illustrated with an extreme. Imagine if the timeout for our data store isn't as
+low as 200ms, but is actually 10s. For example, you might have a relational data
+store where for some customers, 10s queries are (unfortunately) legitimate.
+Reducing the time of worst case queries requires a lot of effort . Dropping the
+query suddenly could potentially leave some customers unable to access certain
+functionality. High timeouts are especially important in a non-threaded
```

Comment:
I don't understand what this last sentence means - why are high timeouts important in a single threaded environment?